support-frontend/assets/components/checkoutBenefits/checkoutBenefitsListContainer.tsx

```diff
@@ -106,9 +30,20 @@ function getBenefitsListTitle(
 	return `For ${priceString} per ${billingPeriod}, you’ll unlock`;
 }
 
+const getbuttonCopy = (
+	higherTier: boolean,
+	thresholdPriceWithCurrency: string,
+	selectedAmount: number,
+) =>
+	higherTier || Number.isNaN(selectedAmount)
+		? null
+		: `Switch to ${thresholdPriceWithCurrency} to unlock all extras`;
+
 export function CheckoutBenefitsListContainer({
 	renderBenefitsList,
```

Comment:
I'm wondering why this `CheckoutBenefitsListContainer` component follows this pattern of taking a function (`renderBenefitsList`) that returns a rendered `CheckoutBenefitsList` component? Is this a common pattern we use? I've not seen it elsewhere, and it's not clear what benefits it offers?

It feels like it'd be cleaner if we replaced the `<CheckoutBenefitsListContainer>` in the [amountAndBenefits.tsx](https://github.com/guardian/support-frontend/blob/df6d7a7889b8d316197f49d41dde64d79feef802/support-frontend/assets/pages/supporter-plus-landing/formSections/amountAndBenefits.tsx)  with`<CheckoutBenefitsList>` and to moved all the logic that gets the values for the `benefitsListProps` out of the `CheckoutBenefitsListContainer` and into the `CheckoutBenefitsList` component, where it's used? We could then get rid of the `CheckoutBenefitsListContainer` which seems like an unnecessary component.

Reply:
This suggestion above ☝️ would mean  `CheckoutBenefitsList` would not take any props, as the logic to generate the `benefitsListProps` would sit within the `checkoutBenefitsList.tsx` module. Would this make it harder to render in StoryBook? Is this why we pass in props to it via a `CheckoutBenefitsListContainer`?

Reply:
This is a presentational and container component approach- the benefit of it is that we separate the logic of talking to Redux, doing calculations, sending tracking, etc etc from the purely presentational UI component. This means:

- It's much easier to write stories for purely presentational components, and to demonstrate (and this have regression tests for) their different variants in Storybook
- It's much easier to swap out one presentational component for a different one- or test two variations of a component against each other- without having to rework any of the underlying business logic

I suggested this as an approach to move towards way back at the start of the year when we were more focused on ways to unify the checkouts, but it will also make A/B testing much less burdensome and generally make our components smaller. I also wrote something up for it on the wiki [here](https://github.com/guardian/support-frontend/wiki/Presentational-and-Container-Components).

Reply:
Ah OK thanks for explaining @i-hardy, this makes sense, and I did suspect my suggestion would make it harder to test after I posted my first message! So if we retain the container `<CheckoutBenefitsListContainer>` component and continue to do all the  talking to Redux, calculations, tracking logic in this component, could we not just return the presentational `<CheckoutBenefitsList>` component rather than the result of the `renderBenefitsList` function? Abridged example...

```
export function CheckoutBenefitsListContainer() {
   const benefitsListProps = { ... } // do all calculations to generate benefitsListProps here
   return <CheckoutBenefitsList {...benefitsListProps} />
}
```

Reply:
We certainly could do that, but the benefit of using a render prop is that it's up to the consumer of `CheckoutBenefitsListContainer` to decide how to use those props, rather than tying this container specifically to that presentational component. We could swap out `CheckoutBenefitsList` for anything else without having to touch this component.

Reply:
If it's helpful, here's where the component is actually being used (along with several other container/presentational components) https://github.com/guardian/support-frontend/blob/main/support-frontend/assets/pages/supporter-plus-landing/formSections/amountAndBenefits.tsx

Reply:
OK yes I see, it's the first time i've seen the container/presentational component pattern in action, so thanks for explanation!